mava/systems/sable/anakin/rec_sable.py

```diff
@@ -0,0 +1,693 @@
+                opt_states,
+                traj_batch,
+                advantages,
+                targets,
+                key,
+                prev_hstates,
+            )
+            return update_state, loss_info
+
+        update_state = (
+            params,
+            opt_states,
+            traj_batch,
+            advantages,
+            targets,
+            key,
+            prev_hstates,
+        )
+
+        # UPDATE EPOCHS
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config.system.ppo_epochs
+        )
+
+        params, opt_states, traj_batch, advantages, targets, key, _ = update_state
+        learner_state = LearnerState(
+            params,
+            opt_states,
+            key,
+            env_state,
+            last_timestep,
+            updated_hstates,
+        )
+        metric = traj_batch.info
+        return learner_state, (metric, loss_info)
+
+    def learner_fn(learner_state: LearnerState) -> ExperimentOutput[LearnerState]:
+        """Learner function.
+
+        This function represents the learner, it updates the network parameters
+        by iteratively applying the `_update_step` function for a fixed number of
+        updates. The `_update_step` function is vectorized over a batch of inputs.
+
+        Args:
+        ----
+            learner_state (NamedTuple):
+                - params (FrozenDict): The initial model parameters.
+                - opt_state (OptState): The initial optimizer state.
+                - key (chex.PRNGKey): The random number generator state.
+                - env_state (LogEnvState): The environment state.
+                - timesteps (TimeStep): The initial timestep in the initial trajectory.
+                - hstates (HiddenStates): The initial hidden states of the network.
+
+        """
+        batched_update_step = jax.vmap(_update_step, in_axes=(0, None), axis_name="batch")
+
+        learner_state, (episode_info, loss_info) = jax.lax.scan(
+            batched_update_step, learner_state, None, config.system.num_updates_per_eval
+        )
+        return ExperimentOutput(
+            learner_state=learner_state,
+            episode_metrics=episode_info,
+            train_metrics=loss_info,
+        )
+
+    return learner_fn
+
+
+def learner_setup(
+    env: MarlEnv, keys: chex.Array, config: DictConfig
+) -> Tuple[LearnerFn[LearnerState], Callable, LearnerState]:
+    """Initialise learner_fn, network, optimiser, environment and states."""
+    # Get available TPU cores.
+    n_devices = len(jax.devices())
+
+    # Get number of agents.
+    config.system.num_agents = env.num_agents
+
+    # PRNG keys.
+    key, net_key = keys
+
+    # Get number of agents and actions.
+    action_dim = int(env.action_spec().num_values[0])
+    n_agents = env.action_spec().shape[0]
+    config.system.num_agents = n_agents
+    config.system.num_actions = action_dim
+
+    # Setup memory configs.
```

Comment:
Can we add this kind of note in the config also if it isn't already there
```suggestion
    # Setting the chunksize - smaller chunks save memory at the cost of speed
```